### PR TITLE
Unary and binary duplicate expression elimination

### LINKED ISSUE #3014: Duplicated unary / binary expression ellimination
Many unary duplicated operations could be eliminated. Like:
~~`~(~x)` -> `x`~~
~~`-(-x)` -> `x`~~
`abs(abs(x))` -> `abs(x)`
`ceil(ceil(x))` -> `ceil(x)`
`floor(floor(x))` -> `floor(x)`
`trunc(trunc(x))` -> `trunc(x)`
`nearest(nearest(x))` -> `nearest(x)`
etc

and some binary:
`y - (y - x)` -> `x` _int only_
`(x ^ y) ^ y` -> `x`
`(x | y) | y` -> `x | y`
`(x & y) & y` -> `x & y`
etc

~~I think it should be something general like `ZeroRemover` and call it like `Deduplicator` or `DupRemover` which walk recursively. Thoughts?~~

### PR DESCRIPTION
**Pull request by MaxGraey**
`abs(abs(x))` -> `abs(x)`
`ceil(ceil(x))` -> `ceil(x)`
`floor(floor(x))` -> `floor(x)`
`trunc(trunc(x))` -> `trunc(x)`
`nearest(nearest(x))` -> `nearest(x)`
`eqz(eqz(bool(x)))` -> `bool(x)`
`sext(sext(x))` -> `sext(x)`
`neg(neg(x))` -> `x`

`y - (y - x)` -> `x`
`(x ^ y) ^ y` -> `x`
`(x | y) | y` -> `x | y`
`(x & y) & y` -> `x & y`
`(x % y) % y` -> `x % y`

also reorder some switch cases and other minor refactorings

Close #3014

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed:

```
Invocations so far:
   FuzzExec: 28886
   CompareVMs: 7404
   CheckDeterminism: 2538
   Wasm2JS: 6627
   Asyncify: 6993

ITERATION: 33626
```

**Comment by MaxGraey** [comment]
refuzzed

File: test/passes/optimize-instructions_all-features.wast [hunk 1]
```diff
@@ -4309,6 +4309,10 @@
     (drop (i32.eqz (i32.eqz (i64.eqz (i64.const 1)))))
     (drop (i32.eqz (i32.eqz (i32.ne (local.get $x) (i32.const 2)))))
 
+    (drop (i32.extend8_s (i32.extend8_s (local.get $x))))
+    (drop (i32.extend16_s (i32.extend16_s (local.get $x))))
+
+
```
Comment (tlively) [inline test/passes/optimize-instructions_all-features.wast:4315]:
```suggestion

```

**Review by tlively — APPROVED** [review]
👍

**Comment by MaxGraey** [comment]
Oh, wait I forgot add `RoundFloat32` and `RoundFloat64` to unary as well)

**Comment by MaxGraey** [comment]
@tlively It ready for landing

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -1397,6 +1403,125 @@ struct OptimizeInstructions
...
+          case CeilFloat64:
+          case FloorFloat64:
+          case TruncFloat64:
+          case NearestFloat64: {
+            // unaryOp(unaryOp(x))  ==>   unaryOp(x)
+            return unaryInner;
+          }
+          case ExtendS8Int32:
+          case ExtendS16Int32: {
+            assert(getModule()->features.hasSignExt());
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1430]: is this assertion necessary? the validator handles stuff like this, so unless there is a specific reason here (like we may emit a signext but want to make sure we don't).

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1430]: Similar checks have other post-MVP parts so I just followed this practice.

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -1397,6 +1403,125 @@ struct OptimizeInstructions
...
+              // x | (x | y)  ==>   x | y
+              if (outer->op == Abstract::getBinary(type, Abstract::And) ||
+                  outer->op == Abstract::getBinary(type, Abstract::Or)) {
+                return inner;
+              }
+            }
+            if (ExpressionAnalyzer::equal(inner->right, outer->left)) {
+              // x ^ (y ^ x)  ==>   y
+              if (outer->op == Abstract::getBinary(type, Abstract::Xor)) {
+                return inner->left;
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1473]: this removes `inner->right`, but that may have side effects? (we only checked `outer->left` and `inner-left` for side effects earlier)

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:1473]: (same issue happens more below I think)

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1473]: > this removes inner->right, but that may have side effects? (we only checked outer->left and inner->left 

No, it removes `outer->left` and `inner->right`. `inner->right` and `inner->left` here the same, so if `inner->left` has side effects than `inner->right` also has side effects

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:1473]: oh right, sorry, I misread it...

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1473]: no, problem

**Pull request merged**